torch/utils/data/dataloader.py

```diff
@@ -1489,17 +1497,25 @@ def _try_put_index(self):
         for _ in range(self._num_workers):  # find the next active worker, if any
             worker_queue_idx = next(self._worker_queue_idx_cycle)
             if self._workers_status[worker_queue_idx]:
-                break
+                if self._in_order:
+                    break
+                elif self._workers_num_tasks[worker_queue_idx] < max_tasks // sum(
+                    self._workers_status
+                ):
+                    # when self._in_order is False, distribute work to a worker if it has capacity
+                    break
         else:
             # not found (i.e., didn't break)
             return
 
         self._index_queues[worker_queue_idx].put((self._send_idx, index))  # type: ignore[possibly-undefined]
         self._task_info[self._send_idx] = (worker_queue_idx,)
+        self._workers_num_tasks[worker_queue_idx] += 1
         self._tasks_outstanding += 1
         self._send_idx += 1
 
-    def _process_data(self, data):
+    def _process_data(self, data, worker_idx):
+        self._workers_num_tasks[worker_idx] -= 1
```

Comment:
I'm pretty sure that the way this is set up, we'll never deadlock due to not decrementing this, but is there somewhere we can give stronger guarantees on this?

Reply:
It _shouldn't_ deadlock the way it is set up, but making that more clear would be helpful.
`_workers_num_tasks` is only incremented in `try_put_index()` which gets called in two places (other than at the very beginning of an epoch to start off a bunch of tasks):
1. `_process_data`, where it is decremented first, so that will be safe, this is also the only way `_next_data` can exit that isn't shutdown/end of epoch
2. within `_next_data` when a worker for an IterableDataset is finished, also safe as that's just distributing work

I think the key thing is that _process_data, and hence the decrementing, is pretty much _always_ called by _next_data when it returns, and that decrementing happens before any incrementing.
I can't think of anything off the top of my head that would give stronger guarantees, but maybe a small bit of refactoring could make this more clear?